File: mobile/ios/Alma/DesignSystem/Sky/SkyField.swift
import SwiftUI

/// A deterministic random source, and the star and mote fields drawn from it.
///
/// **Why seeded rather than `Double.random`.** The sky is regenerated whenever a
/// view's geometry changes — a rotation, a keyboard, a sheet resizing the
/// screen behind it. With an unseeded source every one of those events silently
/// reshuffles the entire star field, which reads as a flicker at exactly the
/// moment the interface is asking for attention elsewhere. With a seed, the same
/// screen has the same sky for as long as the app is running, and two screens
/// with different seeds have visibly different skies.
///
/// SplitMix64 rather than `SystemRandomNumberGenerator`: it is eight lines, it is
/// well-distributed enough for scattering dots, and it has no state we have to
/// think about across threads.
struct SkyRandom: RandomNumberGenerator {
    private var state: UInt64

    init(seed: UInt64) {
        // A zero seed would make SplitMix64 produce a fixed sequence, so it is
        // pushed off zero rather than rejected — a caller passing 0 wants "the
        // default sky", not a crash.
        self.state = seed &+ 0x9E37_79B9_7F4A_7C15
    }

    mutating func next() -> UInt64 {
        state = state &+ 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }

    mutating func unit() -> Double {
        Double(next() >> 11) * (1.0 / 9_007_199_254_740_992.0)
    }

    mutating func between(_ low: Double, _ high: Double) -> Double {
        low + unit() * (high - low)
    }
}

/// One star. Positions are *fractions* of the canvas, not points, so a field
/// generated once survives every layout change without being regenerated.
struct SkyStar: Sendable {
    let x: Double
    let y: Double
    let radius: Double
    /// Where in its own twinkle cycle this star starts, 0...1.
    let phase: Double
    /// Which of the three star colours this one is.
    let tint: Tint

    enum Tint: Sendable {
        /// Almost white — the majority.
        case white
        /// Aged parchment.
        case warm
        /// Gold, and rare: more than a few and the sky starts to look yellow.
        case gold

        var colour: Color {
            switch self {
            case .white: Color(hex: 0xFFF8E6)
            case .warm: Color(hex: 0xE6D9B4)
            case .gold: .almaGold
            }
        }
    }
}

/// One drifting mote of dust.
struct SkyMote: Sendable {
    let x: Double
    /// Where it starts, as a fraction of the canvas height.
    let y: Double
    let size: Double
    /// Seconds for one rise. The web app uses 14; varying it stops the motes
    /// from moving as a group, which is the thing that gives away an animation.
    let duration: Double
    let delay: Double
}

/// A generated sky: the two star layers and the motes, produced once from a
/// seed and then held.
///
/// The counts come straight from the brand book's ceiling — "≤3 dust motes, one
/// comet per screen, at most two aura blobs". The star counts are ours, because
/// the CSS expresses its field as nine hand-placed radial gradients and nine
/// stars on a 900-point canvas is a diagram of a sky rather than a sky.
struct SkyField: Sendable {

    let near: [SkyStar]
    let far: [SkyStar]
    let motes: [SkyMote]

    /// - Parameters:
    ///   - seed: anything stable for the screen. `Route`-derived values are
    ///     ideal; the default is the one every ambient background uses.
    ///   - density: 1.0 is the standard field. The journey's ceremony turns it
    ///     up; a reading screen turns it down so the text is never competing
    ///     with the background.
    init(seed: UInt64 = 0x414C_4D41, density: Double = 1.0) {
        var rng = SkyRandom(seed: seed)

        let nearCount = Int(64 * density)
        let farCount = Int(96 * density)

        func star(minRadius: Double, maxRadius: Double, goldChance: Double) -> SkyStar {
            let roll = rng.unit()
            let tint: SkyStar.Tint = roll < goldChance ? .gold : (roll < 0.34 ? .warm : .white)
            return SkyStar(
                x: rng.unit(),
                y: rng.unit(),
                radius: rng.between(minRadius, maxRadius),
                phase: rng.unit(),
                tint: tint
            )
        }

        // The near layer is the one the eye reads: fewer, larger, brighter.
        self.near = (0..<nearCount).map { _ in
            star(minRadius: 0.7, maxRadius: 1.05, goldChance: 0.10)
        }
        // The far layer is a wash. Uniformly 0.5pt and half-lit, on a slower
        // and offset clock, so the two layers never pulse together.
        self.far = (0..<farCount).map { _ in
            star(minRadius: 0.45, maxRadius: 0.62, goldChance: 0.06)
        }

        self.motes = (0..<3).map { index in
            SkyMote(
                x: rng.between(0.06, 0.94),
                y: rng.between(0.45, 0.95),
                size: rng.between(1.6, 2.4),
                duration: rng.between(12, 17),
                delay: Double(index) * rng.between(3.0, 6.0)
            )
        }
    }
}

File: mobile/flutter/alma/lib/design/sky/sky_field.dart
import 'dart:ui';

import '../palette.dart';

/// Детерминированный источник случайности и поля звёзд, из него вытянутые.
///
/// Порт `mobile/ios/Alma/DesignSystem/Sky/SkyField.swift`.
///
/// **Почему с зерном, а не обычным random.** Небо пересобирается при каждом
/// изменении геометрии — поворот, клавиатура, лист, меняющий размер экрана
/// позади. С незасеянным источником каждое такое событие молча перетасовывает
/// всё звёздное поле, и это читается как мигание ровно в тот момент, когда
/// интерфейс просит внимания в другом месте. С зерном экран весь сеанс имеет
/// одно и то же небо, а два экрана с разными зёрнами — видимо разные.
///
/// SplitMix64, а не встроенный `Random`: восемь строк, распределение достаточно
/// ровное для рассыпания точек, и нет состояния, о котором надо думать между
/// потоками. Dart не имеет беззнакового 64-битного типа, но `int` здесь и есть
/// 64 бита с обёрткой при переполнении, а сдвиг `>>>` беззнаковый — так что
/// последовательность получается **побитово та же, что на iOS**, и небо на двух
/// платформах совпадает звезда в звезду.
class SkyRandom {
  int _state;

  SkyRandom({int seed = 0x414C4D41})
      // Нулевое зерно заставило бы SplitMix64 выдать постоянную
      // последовательность, поэтому оно сдвигается, а не отвергается: тот, кто
      // передал 0, хочет «небо по умолчанию», а не отказ.
      : _state = seed + 0x9E3779B97F4A7C15;

  int next() {
    _state = _state + 0x9E3779B97F4A7C15;
    var z = _state;
    z = (z ^ (z >>> 30)) * 0xBF58476D1CE4E5B9;
    z = (z ^ (z >>> 27)) * 0x94D049BB133111EB;
    return z ^ (z >>> 31);
  }

  double unit() => (next() >>> 11) * (1.0 / 9007199254740992.0);

  double between(double low, double high) => low + unit() * (high - low);
}

/// Оттенок звезды. Их три, и золотых мало намеренно: чуть больше — и небо
/// начинает выглядеть жёлтым.
enum StarTint {
  /// Почти белый — большинство.
  white(Color(0xFFFFF8E6)),

  /// Состаренный пергамент.
  warm(Color(0xFFE6D9B4)),

  /// Золото, и редко.
  gold(AlmaPalette.gold);

  const StarTint(this.colour);
  final Color colour;
}

/// Одна звезда. Положение — **доля** холста, а не точки, чтобы поле, созданное
/// однажды, пережило любое изменение раскладки без пересоздания.
class SkyStar {
  const SkyStar({
    required this.x,
    required this.y,
    required this.radius,
    required this.phase,
    required this.tint,
  });

  final double x;
  final double y;
  final double radius;

  /// С какого места собственного цикла мерцания эта звезда начинает, 0…1.
  final double phase;
  final StarTint tint;
}

/// Одна плывущая пылинка.
class SkyMote {
  const SkyMote({
    required this.x,
    required this.y,
    required this.size,
    required this.duration,
    required this.delay,
  });

  final double x;

  /// Откуда начинает, долей высоты холста.
  final double y;
  final double size;

  /// Секунд на один подъём. Веб берёт 14; разброс не даёт пылинкам двигаться
  /// группой — а именно это и выдаёт анимацию.
  final double duration;
  final double delay;
}

/// Готовое небо: два слоя звёзд и пылинки, произведённые один раз из зерна.
///
/// Количества взяты из потолка бренд-бука — «не больше трёх пылинок, одна
/// комета на экран, не больше двух пятен ауры». Числа звёзд наши: CSS выражает
/// своё поле девятью вручную расставленными градиентами, а девять звёзд на
/// холсте в 900 точек — это схема неба, а не небо.
class SkyField {
  const SkyField._(this.near, this.far, this.motes);

  final List<SkyStar> near;
  final List<SkyStar> far;
  final List<SkyMote> motes;

  /// - [seed]: что угодно устойчивое для экрана. Значение, выведенное из
  ///   маршрута, — идеально; по умолчанию то же, что у любого фонового неба.
  /// - [density]: 1.0 — обычное поле. Церемония путешествия его поднимает;
  ///   экран чтения опускает, чтобы текст ни с чем не соревновался.
  factory SkyField.generate({int seed = 0x414C4D41, double density = 1.0}) {
    final rng = SkyRandom(seed: seed);

    final nearCount = (64 * density).toInt();
    final farCount = (96 * density).toInt();

    SkyStar star({
      required double minRadius,
      required double maxRadius,
      required double goldChance,
    }) {
      final roll = rng.unit();
      final tint = roll < goldChance
          ? StarTint.gold
          : (roll < 0.34 ? StarTint.warm : StarTint.white);
      return SkyStar(
        x: rng.unit(),
        y: rng.unit(),
        radius: rng.between(minRadius, maxRadius),
        phase: rng.unit(),
        tint: tint,
      );
    }

    // Ближний слой — тот, который читает глаз: меньше числом, крупнее, ярче.
    final near = List<SkyStar>.generate(
      nearCount,
      (_) => star(minRadius: 0.7, maxRadius: 1.05, goldChance: 0.10),
    );
    // Дальний слой — заливка. Мельче и вполсилы, на своих, более медленных и
    // сдвинутых часах, чтобы два слоя никогда не пульсировали вместе.
    final far = List<SkyStar>.generate(
      farCount,
      (_) => star(minRadius: 0.45, maxRadius: 0.62, goldChance: 0.06),
    );

    final motes = List<SkyMote>.generate(
      3,
      (index) => SkyMote(
        x: rng.between(0.06, 0.94),
        y: rng.between(0.45, 0.95),
        size: rng.between(1.6, 2.4),
        duration: rng.between(12, 17),
        delay: index * rng.between(3.0, 6.0),
      ),
    );

    return SkyField._(near, far, motes);
  }
}
